File: core/pkg/shamir/shamir.go
package shamir

import (
	"crypto/rand"
	"errors"
	"fmt"
)

var (
	ErrThresholdTooSmall   = errors.New("shamir: threshold K must be at least 2")
	ErrShareCountTooSmall  = errors.New("shamir: share count N must be >= threshold K")
	ErrTooManyShares       = errors.New("shamir: maximum 255 shares (GF(2^8) limit)")
	ErrEmptySecret         = errors.New("shamir: secret must not be empty")
	ErrNotEnoughShares     = errors.New("shamir: need at least 2 shares to reconstruct")
	ErrMismatchedShareLen  = errors.New("shamir: all shares must have the same data length")
	ErrZeroShareIndex      = errors.New("shamir: share index must not be 0")
	ErrDuplicateShareIndex = errors.New("shamir: duplicate share indices")
)

// Share represents a single Shamir share.
type Share struct {
	X byte   // Evaluation point (1..255, never 0)
	Y []byte // Share data (same length as original secret)
}

// Split divides secret into n shares with threshold k.
// Any k shares can reconstruct the secret; k-1 reveal nothing.
func Split(secret []byte, n, k int) ([]Share, error) {
	if k < 2 {
		return nil, ErrThresholdTooSmall
	}
	if n < k {
		return nil, ErrShareCountTooSmall
	}
	if n > 255 {
		return nil, ErrTooManyShares
	}
	if len(secret) == 0 {
		return nil, ErrEmptySecret
	}

	shares := make([]Share, n)
	for i := range shares {
		shares[i] = Share{
			X: byte(i + 1),
			Y: make([]byte, len(secret)),
		}
	}

	// Temporary buffer for polynomial coefficients.
	coeffs := make([]byte, k)
	defer func() {
		for i := range coeffs {
			coeffs[i] = 0
		}
	}()

	for byteIdx := 0; byteIdx < len(secret); byteIdx++ {
		coeffs[0] = secret[byteIdx]
		// Fill degrees 1..k-1 with random bytes.
		if _, err := rand.Read(coeffs[1:]); err != nil {
			return nil, fmt.Errorf("shamir: random generation failed: %w", err)
		}
		for i := range shares {
			shares[i].Y[byteIdx] = evaluatePolynomial(coeffs, shares[i].X)
		}
	}

	return shares, nil
}

// Combine reconstructs the secret from k or more shares via Lagrange interpolation.
func Combine(shares []Share) ([]byte, error) {
	if len(shares) < 2 {
		return nil, ErrNotEnoughShares
	}

	secretLen := len(shares[0].Y)
	seen := make(map[byte]bool, len(shares))
	for _, s := range shares {
		if s.X == 0 {
			return nil, ErrZeroShareIndex
		}
		if len(s.Y) != secretLen {
			return nil, ErrMismatchedShareLen
		}
		if seen[s.X] {
			return nil, ErrDuplicateShareIndex
		}
		seen[s.X] = true
	}

	result := make([]byte, secretLen)
	for byteIdx := 0; byteIdx < secretLen; byteIdx++ {
		var value byte
		for i, si := range shares {
			// Lagrange basis polynomial L_i evaluated at 0:
			// L_i(0) = product over j!=i of (0 - x_j)/(x_i - x_j)
			//         = product over j!=i of x_j / (x_i XOR x_j)
			var basis byte = 1
			for j, sj := range shares {
				if i == j {
					continue
				}
				num := sj.X
				den := Add(si.X, sj.X) // x_i - x_j = x_i XOR x_j in GF(2^8)
				d, err := Div(num, den)
				if err != nil {
					return nil, err
				}
				basis = Mul(basis, d)
			}
			value = Add(value, Mul(si.Y[byteIdx], basis))
		}
		result[byteIdx] = value
	}

	return result, nil
}

// AdaptiveThreshold returns the read threshold K = max(2, floor(n/3)).
// K is the minimum number of shares required to reconstruct the secret; any
// K-1 shares reveal zero information. The floor is 2 (not 3) because Split
// itself requires only K>=2, and pairing K=2 with WriteQuorum's W>K invariant
// gives small clusters (e.g. N=3) read fault tolerance while keeping a
// successful write recoverable.
func AdaptiveThreshold(n int) int {
	t := n / 3
	if t < 2 {
		return 2
	}
	return t
}

// WriteQuorum returns the write quorum W = min(n, max(K+1, ceil(2n/3))).
//
// The invariant W > K is the durability guarantee: any push reported successful
// has persisted strictly more shares than a read requires, so a "successful"
// write is always recoverable and survives the loss of at least one guardian.
// The previous formula (ceil(2n/3), with K floored at 3) produced W=2 < K=3 at
// n=3, so a push could report success while storing too few shares to ever read
// back — silent, permanent data loss.
func WriteQuorum(n int) int {
	if n <= 0 {
		return 0
	}
	k := AdaptiveThreshold(n)
	w := (2*n + 2) / 3 // ceil(2n/3)
	if w < k+1 {
		w = k + 1
	}
	if w > n {
		w = n
	}
	return w
}

// evaluatePolynomial evaluates p(x) = coeffs[0] + coeffs[1]*x + ... using Horner's method.
func evaluatePolynomial(coeffs []byte, x byte) byte {
	var result byte
	for i := len(coeffs) - 1; i >= 0; i-- {
		result = Add(Mul(result, x), coeffs[i])
	}
	return result
}

File: sdk-vault/src/crypto/shamir.ts
/**
 * Shamir's Secret Sharing over GF(2^8)
 *
 * Information-theoretic secret splitting: any K shares reconstruct the secret,
 * K-1 shares reveal zero information.
 *
 * Uses GF(2^8) with irreducible polynomial x^8 + x^4 + x^3 + x + 1 (0x11B),
 * same as AES. This is the standard choice for byte-level SSS.
 */

import { randomBytes } from '@noble/ciphers/webcrypto';

// ── GF(2^8) Arithmetic ─────────────────────────────────────────────────────

const IRREDUCIBLE = 0x11b;

/** Exponential table: exp[log[a] + log[b]] = a * b */
const EXP_TABLE = new Uint8Array(512);

/** Logarithm table: log[a] for a in 1..255 (log[0] is undefined) */
const LOG_TABLE = new Uint8Array(256);

// Build log/exp tables using generator 3
(function buildTables() {
  let x = 1;
  for (let i = 0; i < 255; i++) {
    EXP_TABLE[i] = x;
    LOG_TABLE[x] = i;
    x = x ^ (x << 1); // multiply by generator (3 is primitive in this field)
    if (x >= 256) x ^= IRREDUCIBLE;
  }
  // Extend exp table for easy modular arithmetic (avoid mod 255)
  for (let i = 255; i < 512; i++) {
    EXP_TABLE[i] = EXP_TABLE[i - 255]!;
  }
})();

/** GF(2^8) addition: XOR */
function gfAdd(a: number, b: number): number {
  return a ^ b;
}

/** GF(2^8) multiplication via log/exp tables */
function gfMul(a: number, b: number): number {
  if (a === 0 || b === 0) return 0;
  return EXP_TABLE[LOG_TABLE[a]! + LOG_TABLE[b]!]!;
}

/** GF(2^8) multiplicative inverse */
function gfInv(a: number): number {
  if (a === 0) throw new Error('GF(2^8): division by zero');
  return EXP_TABLE[255 - LOG_TABLE[a]!]!;
}

/** GF(2^8) division: a / b */
function gfDiv(a: number, b: number): number {
  if (b === 0) throw new Error('GF(2^8): division by zero');
  if (a === 0) return 0;
  return EXP_TABLE[(LOG_TABLE[a]! - LOG_TABLE[b]! + 255) % 255]!;
}

// ── Share Type ──────────────────────────────────────────────────────────────

/** A single Shamir share */
export interface Share {
  /** Share index (1..N, never 0) */
  x: number;
  /** Share data (same length as secret) */
  y: Uint8Array;
}

// ── Split ───────────────────────────────────────────────────────────────────

/**
 * Splits a secret into N shares with threshold K.
 *
 * @param secret - Secret bytes to split (any length)
 * @param n - Total number of shares to create (2..255)
 * @param k - Minimum shares needed for reconstruction (2..n)
 * @returns Array of N shares
 */
export function split(secret: Uint8Array, n: number, k: number): Share[] {
  if (k < 2) throw new Error('Threshold K must be at least 2');
  if (n < k) throw new Error('Share count N must be >= threshold K');
  if (n > 255) throw new Error('Maximum 255 shares (GF(2^8) limit)');
  if (secret.length === 0) throw new Error('Secret must not be empty');

  const coefficients = new Array<Uint8Array>(secret.length);
  for (let i = 0; i < secret.length; i++) {
    const poly = new Uint8Array(k);
    poly[0] = secret[i]!;
    const rand = randomBytes(k - 1);
    poly.set(rand, 1);
    coefficients[i] = poly;
  }

  const shares: Share[] = [];
  for (let xi = 1; xi <= n; xi++) {
    const y = new Uint8Array(secret.length);
    for (let byteIdx = 0; byteIdx < secret.length; byteIdx++) {
      y[byteIdx] = evaluatePolynomial(coefficients[byteIdx]!, xi);
    }
    shares.push({ x: xi, y });
  }

  for (const poly of coefficients) {
    poly.fill(0);
  }

  return shares;
}

function evaluatePolynomial(coeffs: Uint8Array, x: number): number {
  let result = 0;
  for (let i = coeffs.length - 1; i >= 0; i--) {
    result = gfAdd(gfMul(result, x), coeffs[i]!);
  }
  return result;
}

// ── Combine ─────────────────────────────────────────────────────────────────

/**
 * Reconstructs a secret from K or more shares using Lagrange interpolation.
 *
 * @param shares - Array of K or more shares (must all have same y.length)
 * @returns Reconstructed secret
 */
export function combine(shares: Share[]): Uint8Array {
  if (shares.length < 2) throw new Error('Need at least 2 shares');

  const secretLength = shares[0]!.y.length;
  for (const share of shares) {
    if (share.y.length !== secretLength) {
      throw new Error('All shares must have the same data length');
    }
    if (share.x === 0) {
      throw new Error('Share index must not be 0');
    }
  }

  const xValues = new Set(shares.map(s => s.x));
  if (xValues.size !== shares.length) {
    throw new Error('Duplicate share indices');
  }

  const secret = new Uint8Array(secretLength);

  for (let byteIdx = 0; byteIdx < secretLength; byteIdx++) {
    let value = 0;

    for (let i = 0; i < shares.length; i++) {
      const xi = shares[i]!.x;
      const yi = shares[i]!.y[byteIdx]!;

      let basis = 1;
      for (let j = 0; j < shares.length; j++) {
        if (i === j) continue;
        const xj = shares[j]!.x;
        basis = gfMul(basis, gfDiv(xj, gfAdd(xi, xj)));
      }

      value = gfAdd(value, gfMul(yi, basis));
    }

    secret[byteIdx] = value;
  }

  return secret;
}

/** @internal Exported for cross-platform test vector validation */
export const _gf = { add: gfAdd, mul: gfMul, inv: gfInv, div: gfDiv, EXP_TABLE, LOG_TABLE } as const;
